src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.Windows.cs

```diff
@@ -299,8 +308,64 @@ public ApartmentState GetApartmentState()
             }
         }
 
-        // TODO: https://github.com/dotnet/corefx/issues/20766
-        public bool TrySetApartmentState(ApartmentState state) { throw new PlatformNotSupportedException(); }
+        public bool TrySetApartmentState(ApartmentState state)
+        {
+            if (this != CurrentThread)
+            {
+                if (HasStarted())
+                    throw new ThreadStateException();
+                _initialAppartmentState = state;
+                return true;
+            }
+
+            if (state != ApartmentState.Unknown)
+            {
+                InitializeCom(state);
+            }
+            else
+            {
+                UninitializeCom();
+            }
+
+            // Clear the cache and check whether new state matches the desired state
+            t_apartmentType = ApartmentType.Unknown;
+            return state == GetApartmentState();
+        }
+
+        internal static void InitializeCom(ApartmentState state = ApartmentState.MTA)
+        {
+            if (t_comInitialized)
+                return;
+
+#if ENABLE_WINRT
+            int hr = Interop.WinRT.RoInitialize(IntPtr.Zero,
+                (state == ApartmentState.STA) ? Interop.WinRT.RO_INIT_SINGLETHREADED
+                    : Interop.WinRT.RO_INIT_MULTITHREADED);
+#else
+            int hr = Interop.Ole32.CoInitializeEx(IntPtr.Zero,
+                (state == ApartmentState.STA) ? Interop.Ole32.COINIT_APARTMENTTHREADED
+                    : Interop.Ole32.COINIT_MULTITHREADED);
+#endif
+            if ((hr < 0) && (hr != HResults.RPC_E_CHANGED_MODE))
+                throw new OutOfMemoryException();
```

Comment:
It looks like RoInitialize could also return E_UNEXPECTED. Would Marshal.ThrowExceptionForHR make sense here?

Reply:
I think at present `E_UNEXPECTED` may be returned in two exotic cases only:
- `CoInitializeEx` is called from the RPCSS process, i.e., svchost.exe running Remote Procedure Call Service.
- `CoInitializeEx` is called from a COM apartment shutdown callback.

Reply:
I will leave it as is. We make simplifying assumption to just throw OOM in several of other places. It is not clear whether trying to call ThrowExceptionForHR where this may be used early startup would work well.